### DISCUSSION TITLE
New Keymap API

### DISCUSSION THREAD

**Discussion by mehalter** [AstroNvim/AstroNvim#679] [Announcements]
We just merged PR #678 which adds a new user configuration API for configuring keymaps both in the general case as well as enabling configuration of the LSP keymaps that we assign in the `on_attach`. This is a similar table format to what is currently used in `which-key.register_mappings`, and looks something like this:

```lua
return {
  lsp = {
    -- override on_attach mappings here
    mappings = {
      -- mode
      n = {
        -- bindings for given mode
        ["<leader>lf"] = false, -- set a mappings to false to disable it
        -- set a new bindings
        ["<leader>lF"] = {
          function()
            vim.lsp.buf.formatting_sync()
          end,
          desc = "Format code",
        },
      },
    },
  },
  -- general mappings configuration
  mappings = {
    n = {
      ["<C-s>"] = ":w!<cr>", -- set to a string or a function to do a basic binding with no options/description
      ["<"] = { "<gv", desc = "unindent line", silent = false, remap = false }, -- add any keymap options as well
    },
    t = {
      ["<esc>"] = false, -- easily disable default bindings
    },
  },
}
```

**Comment by pkazmier** [comment]
As a new user, I'm somewhat confused about when to use each of the available means to setup mappings: the `mappings` table, `lsp.mappings` table, and the `which-key` plug-in configuration. Do you have any guidance for new users? Much appreciated!

**Reply by mehalter** [reply]
@pkazmier the mappings table can definitely do similar things to the `which-key` table. The `which-key` table has a much more nested structure designed for creating nested menus. This does not get covered by the mappings table. You can just use the `mappings` table for all your mapping needs and ignore the `which-key` plugin configuration, but the mappings table will not allow you to put menu titles to nested menus in which-key so we can't just drop one.

**Comment by meijieru** [comment]
How to disable the key from `astroncommunity` like [it](https://github.com/AstroNvim/astrocommunity/blob/ba3ea511fbc3fa8113485859681ce5fe3c2d6d4e/lua/astrocommunity/file-explorer/oil-nvim/init.lua#L7)?
